File: apps/ex_marketer_web/assets/js/components/channel/userChannel.js
import socket from "../socket";

class UserChannel {
  constructor() {
    this.channel = socket.channel(`user:${window.userId}`, {});

    this._setup();
  }

  _setup() {
    this.channel
      .join()
      .receive("ok", (resp) => {
        this.channel.on("keyword_successed", (message) => {
          console.log(message);
        });

        this.channel.on("keyword_fail", (message) => {
          console.log(message);
        });
      })
      .receive("error", (resp) => {
        throw resp;
      });
  }
}

export default UserChannel;

File: apps/ex_marketer_web/lib/ex_marketer_web/channels/user_channel.ex
defmodule ExMarketerWeb.UserChannel do
  use ExMarketerWeb, :channel

  alias ExMarketer.Keyword
  alias Phoenix.View
  alias ExMarketerWeb.KeywordView

  intercept ["keyword_completed"]

  def join("user:" <> id, _params, socket) do
    if String.to_integer(id) === socket.assigns[:current_user_id] do
      {:ok, socket}
    else
      {:error, :unauthorized}
    end
  end

  # ExMarketerWeb.Endpoint.broadcast!
  def handle_out("keyword_completed", params, socket) do
    keyword_id = params.keyword_id
    keyword = Keyword.find(keyword_id)

    if socket.assigns[:current_user_id] === keyword.user_id do
      push(socket, "keyword_completed", %{
        keyword_id: keyword_id,
        keyword_view: build_keyword_view(keyword)
      })
    end

    {:noreply, socket}
  end

  defp build_keyword_view(keyword) do
    View.render_to_string(KeywordView, "_keyword.html", %{keyword: keyword, recently_update: true})
  end
end
